compiler/vm/cbackend2.nim

```diff
@@ -0,0 +1,1097 @@
+
+    # XXX: the current approach has the downside that code which may end up
+    #      not being part of the alive graph is still put through all
+    #      processing. A different approach would be to run all processing
+    #      (dependency collection/irgen and the IR passes) in a loop until no
+    #      more new dependencies are found
+
+    # run ``irgen`` for compilerprocs and their (not yet seen) dependencies
+    for it in g.compilerprocs.items:
+      if it.kind in routineKinds:
+        nextProcs.add it
+
+    drain(c, g.config, env, procImpls, nextProcs, nextProcs2, seenProcs)
+
+  # generate all deferred symbols and declarations. ``c.defSyms`` should not
+  # be used past this point
+  flush(c.defSyms, g.cache, env.syms)
+
+  # XXX: it would be better to also set the type during
+  #      ``DeferredSymbols.flush``. But then the procedure needs access to a
+  #      ``DeferredTypeGen`` object - which is a bit awkward, since
+  #      ``DeferredTypeGen.flush`` requires a ``DeferredSymbol`` object
+  for id, s in env.syms.msymbols:
+    if (let orig = env.syms.orig.getOrDefault(id); orig != nil):
+      s.typ = c.types.requestType(orig.typ)
+
+  # XXX: this is a problem. Resolving the type-bounds operation requires
+  #      all alive types being present, but that isn't the case yet since
+  #      the final flush hasn't happenend, but for that we need to call
+  #      ``finish``, but we can't since we're still adding procedures
+  resolveTypeBoundOps(passEnv, g, c.types, c.procs)
+
+  block:
+    # translate the literal data from it's ``PNode``-based representation
+    for id, data in c.constData.pairs:
+      assert env.syms[id].kind == skConst
+      env.syms.setData(id): add(c.data, c.procs, g.config, data)
+
+    reset c.constData # no longer needed
+
+  c.procs.finish(c.types, g.cache)
+
+  block:
+    # flush all remaining deferred types. ``c.symEnv`` was consumed when
+    # flushing it, so we create a temporary new ``DeferredSymbol`` object here
+    var defSyms = initDeferredSyms(env.syms)
+    c.types.flush(env.types, defSyms, g.config)
+    c.types.flush2(env.types, c.data, g.config)
+
+    flush(defSyms, g.cache, env.syms)
+
+  # replace all placeholder type IDs
+  finishTypes(passEnv, c.types, procImpls, c.procs, env.syms)
+
+  processObjects(passEnv, g.cache, env.types, env.syms, c.types.objects)
+
+  swap(c.procs, env.procs)
+  swap(c.data, env.data)
+
+  let objects = move c.types.objects
+
+  # TODO: ``generateCode`` needs to be split up. Freeing the ``TCtx`` should
+  #       happen automatically as part of stack-frame clean up
+  reset(c) # we no longer need the irgen context
+
+  block:
+    # all alive globals are collected by now - register them with their
+    # owning module
+
+    for id in env.syms.items:
+      let s = env.syms[id]
+      case s.kind
+      of skVar, skLet, skForVar:
+        let mIdx = mlist[].lookupModule(env.syms.orig[id])
+        modules[mIdx].syms.add(id)
+
+      else:
```

Comment:
To make sure I understand correctly: if a `case|if` branch doesn't end in a nested statement/expression, it's preferred to not use a new line?

Reply:
I'm still working through what's tasteful, but that's not a bad way to look at it.

The only time I find separating it out helps is when there are many branches in a single `case` and they don't fit the screen vertical space. An example of this would be the report body printing in cli_reporter.

Sorry this is still a bit fuzzy, I'm refining it as fast as I can. My mental model is, the eye is rapidly skimming down the lines, a blank line is a big interrupt to that, so the conceptual separation between the contiguous lines before and the lines thereafter needs to be very meaningful. Often times with conditional branches they're not really conceptually separate. They are in that report body scenario, because you can mentally elide every other branch and the proc body reads like a specialized print for that one case.